---
title: "EPID 600 Project Template"
author: "Your Name"
output: html_document
---

Use this template to complete your project throughout the course. Your Final Project presentation in class will be based on the contents of this document. Due dates:

- Topic choice: 9/18/15
- List of faculty/staff to be contacted: 10/2/15
- Complete overview with data source to address a specific problem identified: 10/16/15
- Introduction: 10/30/15
- Methods and Results draft: 11/13/15
- Final Document: 11/30/15

### Overview
In this section, give a brief a description of your project and its goal, what data you are using to complete it, and what three faculty/staff in different fields you have spoken to about your project with a brief summary of what you learned from each person. 


### Introduction 
In the first paragraph, describe the problem addressed, its significance, and some background to motivate the problem.

In the second paragraph, explain why your problem is interdisciplinary, what fields can contribute to its understanding, and incorporate background related to what you learned from meeting with faculty/staff.

Title: Predicting Warfarin Dosage from clinical and pharmacogenetic data

The formation of blood clots in the blood vessels and their migration else where in the body can be prevented by warfarin, an anticoagulant. More than 30 million patients are prescribed warfarin, however, the proper dose of warfarin is difficult to determine because it can vary amongnpatients by a factor of 10. The consequences of taking the incorrect dose can be fatal. The first dose administered is usually 5 mg and the time it takes for the blood to clot is measured. This is called Prothrombin Time(PT) or the International Normalized Ratio(INR). INR is the same as PT but normalized based on a person of the same demographic. The INR value should be prolonged by the warfarin, and based on this measure, the next daily dose will be adjusted. The patient has to stay in the hospital until the INR value is in the normal range for at least 2 repetitions. Studies have shown that warfarin activity can be determined partially by genetic factors. Specifically, the two variants in VKORC1 and CYP2C9, which are genes involved in warfarin response. 

The goal of this project is to predict therapeutic warfarin dosage by applying machine learning approaches to a large publicly avaiable data that integrates clincal information as well as genotypic data. The data consists of 5,700 patients and 67 features and by utilizing computational and statistical methods to successfully predict warfarin dosage, the cost and time of adminstration of warfarin dosage will reduce significantly. 

This project is highly interdisciplinary, as it combined the field of pharmacology, genomics, computer science, statistics, and health care. Integrating knowledge from these disciplines will contribute to the field of precision medicine and will change how health care is practiced by improving the prognosis, diagnosis, and treatment of patients.

### Methods
In the first paragraph, describe the data used and general methodological approach. Subsequently, incorporate full R code necessary to retrieve and clean data, and perform analysis. Be sure to include a description of code so that others (including your future self) can understand what you are doing and why.


The International Warfarin Pharmacogenomic Consortium comprised of 21 research groups that have contributed clinical and genetic data for 5,700 subjects who were treated with warfarin. Clinical factors that have been previously associated with warfarin therapy and easily accessible were chosen. The data consists of features regarding demographic information, primary indication of warfarin treatment, the stable therapeutic dose of warfarin, the INR achieved with a stable warfarin dose, the desired INR, the use of concomitant medications, the presence of variants of CYP2C9 AND VKORC1. There is also information regarding race, whether or not the patient is a current smoker, and intake of herbal medication, vitamins, and supplement. However, the missing-ness rate was very large, so it was removed. The outcome variable was the stable therapeutic dose of warfarin, defined as the steady-state dose that led to stable anticoagulation levels.

The data was tested for missinginess rate for features and individuals, in order to see if any variable or subject should be thrown out for lack of succificent data. The initial total of 63 variables were reduced to 23 variables, and individuals that did not have data for all of the 23 variables were thrown out to work on a completed dataset. Linear regression was performed individuall for each feature, to determine if there are any features that are significantly associated with therapeutic dosage. These variables will be then chosen for a model to train through cross-validation. Different methods/models are currently being explored.

Code used to clean and explore data is currently in project_warfarin.R.
```{r eval = TRUE}

#LOAD LIBRARIES
library(RCurl)
library(dplyr)
library(ggplot2)

#CLEAN DATA

#READ IN DATA FROM DATASET IN REPOSITORY
my.url <- getURL("https://raw.githubusercontent.com/misraj/pharmgkb_project/master/warfarin_data.txt") #Getting warfarin data
warfarin <- read.delim(text = my.url, header=TRUE)

#Filter out observations that did not have a thereapeutic dosage data available
warfarin <- filter(warfarin, complete.cases(Therapeutic.Dose))

#filter out observations whose therapeutic dosage  mg/week is < 70, because usually the maximum dosage is 10 mg a day. Being able to detect risks for individuals who need higher dosages is a much more complicated task
warfarin <- filter(warfarin, Therapeutic.Dose <= 70)
#Plot dosage data
N = length(warfarin$Subject.ID)
V = length(names(warfarin))
plot(seq(1:N), warfarin$Therapeutic.Dose, xlab  = "Subject", ylab = "Therapeutic Warfarin Dosage", pch = ".")

# To make this problem a bit more solvable, I divided the total dosage/week by 7 to get an average daily dose. 
warfarin$Therapeutic.Dose <- as.integer(warfarin$Therapeutic.Dose/7)
plot(seq(1:N), warfarin$Therapeutic.Dose, xlab  = "Subject", ylab = "Therapeutic Warfarin Dosage", pch = ".")


#Remove Variables that give us 0 information. No data reported for this field.
warfarin<-dplyr::select(warfarin,-(Cerivastatin))

```

```{r eval = FALSE}
#Look at individual univariate associations
uni_reg <- data.frame(seq(1:V-2),seq(1:V-2))
colnames(uni_reg) <- c("Variable", "Pval")
#First two fields are sample and subject Id, so ignore for univarate analysis
for( i in 3:V){
  datas.glm <- lm(Therapeutic.Dose~ warfarin[,i:i], data=warfarin)
  pv <-summary(datas.glm)$coefficients[2,4]
  uni_reg[i-2,1] <-names(warfarin)[i]
  uni_reg[i-2,2] <-pv
}

#Sort by most significantly associated
uni_reg <- arrange(uni_reg,Pval)
significant <- filter(uni_reg, Pval < 0.05)
significant

# About 41 variables independently are significantly associated with therapeutic dosage
#But we cannot do a linear regression model if there are missing data so we need to filter the data more

#Looking at missingness rate
missingness_feature <- rep(0, V)
for(i in 1:V){
  table <-table(is.na(warfarin[,i]))
  if(length(table) == 1){
    missingness_feature[i] = table/N
  }
  else{
    missingness_feature[i] = table[1]/N
  }
}
miss_feature<- data.frame(names(warfarin),missingness_feature)
miss_feature<- arrange(miss_feature, desc(missingness_feature))
#How many patients have this information?
set_n <- miss_feature$missingness_feature * N
miss_feature <- data.frame(miss_feature, set_n)

#Check missingness of patients
missingness_subjects <- rep(0,N)
for(i in 1:N){
  table <- table(is.na(warfarin[i,]))
  if(length(table) == 1){
    missingness_subjects[i] = table/V
  }
  else{
    missingness_subjects[i] = table[1]/V
  }
}
miss_subject <- data.frame(rownames(warfarin),missingness_subjects)
miss_subject <- arrange(miss_subject, desc(missingness_subjects))

#No patients should be thrown out.
```





```{r eval = TRUE}
#Filter words based on the combined list of missingness and significant univariate associations
warfarin1 <- filter(warfarin, complete.cases(Weight)) 
warfarin1 <- filter(warfarin1, complete.cases(Height))
warfarin1 <- filter(warfarin1, complete.cases(Race.OMB))
warfarin1 <- filter(warfarin1, complete.cases(VKORC1.4_consensus)) #2718
warfarin1 <- filter(warfarin1, complete.cases(VKORC1.1_consensus)) #2208
warfarin1 <- filter(warfarin1, complete.cases(VKORC1.3_consensus))
warfarin1 <- filter(warfarin1, complete.cases(INR.Reported))
warfarin1 <- filter(warfarin1, complete.cases(VKORC1.6_consensus))
warfarin1 <- filter(warfarin1, complete.cases(Site))
warfarin1 <- filter(warfarin1, complete.cases(VKORC1.5))
warfarin1 <- filter(warfarin1, complete.cases(Amiodarone))
warfarin1 <- filter(warfarin1, complete.cases(Race.Reported))
warfarin1 <- filter(warfarin1, complete.cases(Gender))
warfarin1 <- filter(warfarin1, complete.cases(Genotyped.QC3.Cyp2C9))
warfarin1 <- filter(warfarin1, complete.cases(Genotyped.QC2.Cyp2C9))
warfarin1 <- filter(warfarin1, complete.cases(CYP2C9_consensus))
warfarin1 <- filter(warfarin1, complete.cases(VKORC1.7_QC))

bmi <- rep(0,length(warfarin1$Weight))
for(i in 1:length(warfarin1$Weight)){
  h <- warfarin1$Height[i]/100
  den <- h *h
  bmi[i] <- warfarin1$Weight[i]/ den 
  if(bmi[i] < 18.5){
    bmi[i] <- 1
  }
  else if(bmi[i] >=18.5 && bmi[i] <24.9 ){
    bmi[i] <- 2
  }
  else if(bmi[i] >= 25.0 && bmi[i] < 29.9 ){
    bmi[i] <- 3
  }
  else if(bmi[i] >=30.0 && bmi[i] <34.9 ){
    bmi[i] <- 4
  }
  else if(bmi[i] >=35.0 && bmi[i] <39.9 ){
    bmi[i] <- 5
  }
  else{
    bmi[i] <- 6
  }
}

warfarin1 <- cbind(warfarin1, bmi)
warfarin1 <- mutate(warfarin1, bmi = factor(bmi, levels=c(1, 2,3,4,5,6), labels=c("underweight", "normal", "overweight", "obesity1", "obesity2", "obesity3")))

warfarin1$Site = as.factor(warfarin1$Site)
warfarin1$Weight = as.factor(warfarin1$Weight)
warfarin1$Height = as.factor(warfarin1$Height)
warfarin1$INR.Reported = as.factor(warfarin1$INR.Reported)

warfarin1 <- dplyr::select(warfarin1, Subject.ID, Sample.ID, Site, Race.OMB,Race.Reported, Gender, Weight, Height,bmi, VKORC1.4_consensus,VKORC1.1_consensus,VKORC1.3_consensus,INR.Reported,VKORC1.6_consensus, VKORC1.5, Amiodarone, Combined.QC4.CYP2C9,  Genotyped.QC3.Cyp2C9, Genotyped.QC3.Cyp2C9, Reached.Stable.Dose, VKORC1.1_QC,VKORC1.2_QC,VKORC1.3_QC,VKORC1.4_QC,VKORC1.5_QC,VKORC1.6_QC,VKORC1.7_QC,Age,Cyp2C9,CYP2C9_consensus, Therapeutic.Dose)


#Mutate data so that there values are all factors
###Mutate/Change values to factors#######
warfarin1 <- mutate(warfarin1, Amiodarone = factor(Amiodarone, levels=c(0, 1), labels=c("no", "yes")))
warfarin1 <- mutate(warfarin1, Reached.Stable.Dose = factor(Reached.Stable.Dose, levels=c(0, 1), labels=c("no", "yes")))


```


```{r eval= TRUE}
#Split data into training set and validation set

#Get total points
n1 = length(warfarin1$Subject.ID)
#Split data into x.train, x.test, y.train, and y.test
train_rows1 <- sample(1:n1, .66*n1)
warf.train1 <- warfarin1[train_rows1, ]
warf.test1 <- warfarin1[-train_rows1, ]
dose.train1 <- warfarin1$Therapeutic.Dose[train_rows1]
dose.test1 <- warfarin1$Therapeutic.Dose[-train_rows1]
#Get rid of y in dataset
warf.train1 <- dplyr::select(warf.train1, -(Therapeutic.Dose))
warf.test1 <- dplyr::select(warf.test1, -(Therapeutic.Dose))
#Get rid of ID, it will mess up stepwise regression: test and train
warf.test1 <- dplyr::select(warf.test1, -(Subject.ID))
warf.train1 <- dplyr::select(warf.train1, -(Subject.ID))
warf.test1 <- dplyr::select(warf.test1, -(Sample.ID))
warf.train1 <- dplyr::select(warf.train1, -(Sample.ID))

null = lm(dose.train1 ~ 1, data = warf.train1)
full = lm(dose.train1 ~ ., data = warf.train1)
model_selection = step(null, scope=list(lower=null, upper=full), direction="both", criterion = "BIC")
```


```{r eval = TRUE}
###EVALUATION###
warf_train1_reg <- lm(dose.train1 ~ VKORC1.3_consensus + Weight + Age + Cyp2C9 + Amiodarone + 
    Genotyped.QC3.Cyp2C9 + Site + Reached.Stable.Dose, data= warf.train1)
summary(warf_train1_reg)
best.warf_train1_reg <- predict.lm(warf_train1_reg,warf.train1, type = "response")
best.warf_train1_reg <-as.integer(best.warf_train1_reg)
compare_warf_train1 <- data.frame(seq(1:length(warf.train1$Weight)),dose.train1,best.warf_train1_reg )
dose1<- arrange(compare_warf_train1, desc(dose.train1))
colnames(dose1)<- c("Subject","True","Predicted")
SSE1 = sum((dose1$Predicted - dose1$True)^2)
err1 = mean(abs(dose1$Predicted - dose1$True))
acc1 = dose1$Predicted - dose1$True
est1 = seq(1:length(dose1$Dose))
for( i in 1:length(acc1)){
  if (acc1[i] == 0){
    est1[i] = "c"
  }
  else if(acc1[i] > 0){
   est1[i] = "o"
  }
  else{
   est1[i] = "u"
  }
}
dose1$Label = est1
ggplot(data=dose1, aes(x= Predicted, y= True,colour = Label)) +
    geom_point( size=1, shape=21, fill="white")

ggplot(data=dose1, aes(x= Label)) + geom_bar()

###EVALUATION ON TESTING SET FOR MODEL 1###
warf_test1_reg <- lm(dose.test1 ~ VKORC1.3_consensus + Weight + Age + Cyp2C9 + Amiodarone + 
    Genotyped.QC3.Cyp2C9 + Site + Reached.Stable.Dose,data = warf.test1)
summary(warf_test1_reg)
best.warf_test1_reg <- predict.lm(warf_test1_reg,warf.test1, type = "response")
best.warf_test1_reg <-as.integer(best.warf_test1_reg)
compare_warf_test1 <- data.frame(seq(1:length(warf.test1$Weight)),dose.test1,best.warf_test1_reg )
dose1.t<- arrange(compare_warf_test1, desc(dose.test1))
colnames(dose1.t)<- c("Subject","True","Predicted")
SSE1t = sum((dose1.t$Predicted - dose1.t$True)^2)
err1t = mean(abs(dose1.t$Predicted - dose1.t$True))
err.sd1t = sd(abs(dose1.t$Predicted - dose1.t$True))
avg_dose_predicted1t = mean(dose1.t$Predicted)
avg_dose_true1t = mean(dose1.t$True)
dose_pred.sd1t = sd(dose1.t$Predicted)
dose_true.sd1t = sd(dose1.t$True)
acc1t = dose1.t$Predicted - dose1.t$True
est1t = seq(1:length(dose1.t$Dose))
for( i in 1:length(acc1t)){
  if (acc1t[i] == 0){
    est1t[i] = "c"
  }
  else if(acc1t[i] > 0){
    est1t[i] = "o"
  }
  else{
    est1t[i] = "u"
  }
}
dose1.t$Label = est1t
ggplot(data=dose1.t, aes(x= Predicted, y= True,colour = Label)) +
  geom_point( size=1, shape=21, fill="white")

ggplot(data=dose1.t, aes(x= Label)) + geom_bar()

```
### Results
Describe your results and include relevant tables, plots, and code/comments used to obtain them. End with a brief conclusion of your findings related to the question you set out to address.

By performing linear regression on each variable, the following features were shown to be signficantly associated with therapeutic dosage of warfarin: Site, Race defined by Office of Management and Budget, Combined separate genotypes for *2 and *3 into single CYP2C9 diplotypes with possible values *1/*1, *1/*2, *1/*3, *2/*2, *2/*3, *3/*3, VKORC1 genotype: -1639 G>A (3673); chr16:31015190; rs9923231; C/T, weight, whether or not they had a valve replacement, whether or not they take aspirin, whether or not they take Simvastatin, VKORC1 QC genotype: -1639 G>A (3673); chr16:31015190; rs9923231; C/T, INR reported, and whether or not they reached stable dose. 


Diabetes